pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/LLCSegmentCompletionHandlers.java

```diff
@@ -276,10 +307,193 @@ public String segmentUpload(@QueryParam(SegmentCompletionProtocol.PARAM_INSTANCE
+            .withMemoryUsedBytes(memoryUsedBytes);
+    LOGGER.info("Processing segmentCommitEnd:{}", requestParams.toString());
+
+
+    final boolean isSuccess = true;
+    final boolean isSplitCommit = true;
+    SegmentMetadataImpl segmentMetadata = extractMetadataFromInput(metadataFiles, segmentName);
+    // If it fails to extract metadata from the input form, try to download the segment and extract it from the segment.
+    if (segmentMetadata == null) {
+      LOGGER.info("Failed to extract segment metadata for {} from input form, fallback to use the segment file.",
+              segmentName);
+      try {
+        segmentMetadata = extractMetadataFromSegmentFile(segmentName, new URI(segmentLocation));
+      } catch (URISyntaxException e) {
+        LOGGER.error("Invalid segment location: ", segmentLocation);
+        return SegmentCompletionProtocol.RESP_FAILED.toJsonString();
+      }
+    }
+    // Return failure to server if both extraction efforts fail.
+    if (segmentMetadata == null) {
+      LOGGER.warn("Segment metadata extraction failure for segment {}", segmentName);
+      return SegmentCompletionProtocol.RESP_FAILED.toJsonString();
+    }
+    SegmentCompletionProtocol.Response response =
+            SegmentCompletionManager.getInstance().segmentCommitEnd(requestParams, isSuccess, isSplitCommit,
+                    CommittingSegmentDescriptor.fromSegmentCompletionReqParamsAndMetadata(requestParams, segmentMetadata));
+    final String responseStr = response.toJsonString();
+    LOGGER.info("Response to segmentCommitEnd:{}", responseStr);
+    return responseStr;
+  }
+
+  /**
+   * Extract and return the segment metadata from the two input form data files (metadata file and creation meta).
+   * Return null if any of the two files is missing or there is exception during parsing and extraction.
+   *
+   */
+  private SegmentMetadataImpl extractMetadataFromInput(FormDataMultiPart metadataFiles, String segmentNameStr) {
+    String tempMetadataDirStr = StringUtil.join("/", _controllerConf.getLocalTempDir(), segmentNameStr + METADATA_TEMP_DIR_SUFFIX);
+    File tempMetadataDir = new File(tempMetadataDirStr);
+
+    try {
+      Preconditions.checkState(tempMetadataDir.mkdirs(), "Failed to create directory: %s", tempMetadataDirStr);
+      // Extract metadata.properties from the metadataFiles.
+      if (!extractMetadataFromInputField(metadataFiles, tempMetadataDirStr, V1Constants.MetadataKeys.METADATA_FILE_NAME)) {
+        return null;
+      }
+      // Extract creation.meta from the metadataFiles.
+      if (!extractMetadataFromInputField(metadataFiles, tempMetadataDirStr, V1Constants.SEGMENT_CREATION_META)) {
+        return null;
+      }
+      // Load segment metadata
+      return new SegmentMetadataImpl(tempMetadataDir);
+    } catch (Exception e) {
+      LOGGER.error("Exception extracting and reading segment metadata for " + segmentNameStr, e);
```

Comment:
Please use {} for logging segmentNamemStr

Reply:
Done.